File: Leetcode/1973.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
using ll = long long;

class Solution {
public:
    unordered_map<TreeNode*, ll> dp;
    int ans;
    ll solve(TreeNode* root) {
        if (dp.find(root) != dp.end()) {
            return dp[root];
        }
        dp[root] = root->val;
        if (root->left) {
            dp[root] += solve(root->left);
        }
        if (root->right) {
            dp[root] += solve(root->right);
        }
        return dp[root];
    }
    void dfs(TreeNode* root) {
        if (root) {
            if (solve(root) - root->val == root->val) {
                ++ans;
            }
            dfs(root->left);
            dfs(root->right);
        }
    }
    int equalToDescendants(TreeNode* root) {
        dp.clear();
        ans = 0;
        dfs(root);
        return ans;
    }
};

File: Leetcode/1973.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var dp map[*TreeNode]int64

func solve(node *TreeNode) int64 {
	val, ok := dp[node]
	if ok {
		return val
	}
    dp[node] = int64((*node).Val)
	if (*node).Left != nil {
		dp[node] += solve((*node).Left)
	}
	if (*node).Right != nil {
		dp[node] += solve((*node).Right)
	}
	return dp[node]
}

func equalToDescendants(root *TreeNode) int {
	if root == nil {
		return 0
	}
	const MAXN int = 1e5+5
	dp = nil
	dp = make(map[*TreeNode]int64)
	queue := make(chan *TreeNode, MAXN)
	// bfs
	queue <- root
	ans := 0
	for ; len(queue) > 0; {
		node := <- queue
		if solve(node) == int64((*node).Val) << 1 {
			ans++
		}
		if (*node).Left != nil {
			queue <- (*node).Left
		}
		if (*node).Right != nil {
			queue <- (*node).Right
		}
	}
	close(queue)
	return ans
}